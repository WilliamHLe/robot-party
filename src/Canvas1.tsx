$(document).ready(function() {
    //Endrer SVG elementenes farge når man trykker på de
    let c = document.getElementById("canvas");
    let ctx = c.getContext("2d");
    //Data for alle figurene som skal tegner i canvas
    let dance = [
        //Head
        { x: 320, y: 175, w: 105, h: 125, rotate: 0 },
        //Left arm
        { x: 365, y: 290, w: 50, h: 100, rotate: -10 },
        {
            x: 385,
            y: 355,
            w: 100,
            h: 50,
            rotate: 10
        },
        //Left foot
        { x: 400, y: 400, w: 60, h: 125, rotate: -40 },
        { x: 425, y: 490, w: 50, h: 100, rotate: 10 },
        //Body
        { x: 375, y: 325, w: 0, h: 0, rotate: 0 },
        { x: 375, y: 375, w: 0, h: 0, rotate: 0 },
        //Right arm
        { x: 345, y: 300, w: 50, h: 100, rotate: -10 },
        {
            x: 365,
            y: 365,
            w: 100,
            h: 50,
            rotate: 10
        },
        //Right foot
        { x: 325, y: 415, w: 60, h: 125, rotate: 5 },
        { x: 315, y: 515, w: 50, h: 100, rotate: 20 }
    ];
    let bop = [
        //Head
        { x1: 375, x2: 450, x3: 300, y1: 275, y2: 150 },
        //Left arm
        { x: 450, y: 280, w: 50, h: 100, rotate: 90 },
        {
            x: 475,
            y: 275,
            w: 100,
            h: 50,
            rotate: -80
        },
        //Right arm
        { x: 255, y: 280, w: 50, h: 100, rotate: 90 },
        {
            x: 175,
            y: 275,
            w: 100,
            h: 50,
            rotate: 80
        },
        //Right foot
        { x: 290, y: 400, w: 60, h: 125, rotate: 40 },
        { x: 265, y: 490, w: 50, h: 100, rotate: -10 },
        //Left foot
        { x: 410, y: 400, w: 60, h: 125, rotate: -40 },
        { x: 435, y: 490, w: 50, h: 100, rotate: 10 },
        //Body
        { x: 375, y: 325, w: 0, h: 0, rotate: 0 },
        { x: 375, y: 375, w: 0, h: 0, rotate: 0 }

    ];
    let circles = [
        {x:375,y:275,r:150,sA:0.75*Math.PI,eA:0.25*Math.PI,color:"rgb(255,0,0)"},
        {x:375,y:450,r:200,sA:0,eA:2*Math.PI,color:"rgb(0,255,0)"},
        {x:450,y:475,r:75,sA:0,eA:2*Math.PI,color:"rgb(0,0,255)"},
        {x:375,y:75,r:25,sA:0,eA:2*Math.PI,color:"rgb(0,0,255)"}

    ]
    drawFloor();
    ctx.fillStyle = "yellow";
    ctx.lineWidth = 5;
    //Tegner figurene
    function drawDancer(arr) {
        arr.forEach(r => {
            ctx.save();
            ctx.beginPath();
            ctx.fillStyle = "yellow";
            if (r.w == 0) {
                ctx.arc(r.x, r.y, 55, 0, 2 * Math.PI);
            } else if (typeof r.x1 !== "undefined") {
                ctx.beginPath();
                ctx.moveTo(r.x1, r.y1);
                ctx.lineTo(r.x2, r.y2);
                ctx.lineTo(r.x3, r.y2);
                ctx.closePath();
                ctx.stroke();
                ctx.fill();
            } else {
                //Kode som gjør av rektangler roterer rundt midtpunket sitt. Koden er funnet her: https://stackoverflow.com/questions/17125632/html5-canvas-rotate-object-without-moving-coordinates/17126036
                ctx.translate(r.x + r.w / 2, r.y + r.h / 2);
                ctx.rotate((r.rotate * Math.PI) / 180);
                ctx.rect(-r.w / 2, -r.h / 2, r.w, r.h);
            }
            ctx.fill();
            ctx.stroke();
            ctx.restore();
        });
    }
    function drawCircles() {
        ctx.beginPath();
        ctx.moveTo(375, 125);
        ctx.lineTo(375, 100);
        ctx.stroke();
        circles.forEach(c => {
            ctx.beginPath();
            ctx.fillStyle = c.color;
            ctx.arc(c.x,c.y,c.r,c.sA,c.eA);
            ctx.fill();
            ctx.stroke();
        })
    }
    //Tegner gulvet
    function drawFloor() {
        ctx.globalAlpha = 1;
        let gradient = ctx.createLinearGradient(0, 450, 0, 750);
        gradient.addColorStop(0, "grey");
        gradient.addColorStop(1, "white");
        ctx.fillStyle = gradient;
        ctx.fillRect(0, 450, 750, 300);
        //ctx.globalAlpha = 0.5;
    }

    let j = 1;
    let forward = true;
    let speed = 30;
    let utstilling = "bop";
    requestAnimationFrame(animate);
    function animate(){
        ctx.clearRect(0, 0, c.width, c.height);
        drawFloor();
        if(utstilling == "dance") {
            drawDancer(dance);
        } else if(utstilling == "blink"){
            drawCircles();
            if(j > speed || j < 1) {
                circles.forEach(c => {
                    let r = Math.floor(Math.random() * 255);
                    let g = Math.floor(Math.random() * 255);
                    let b = Math.floor(Math.random() * 255);
                    c.color = "rgb("+r+","+g+","+b+")";
                })
            }
        } else if(utstilling == "bop") {
            drawDancer(bop);
        }
        if(forward){
            //Utstilling 1
            dance[2].x += 40/speed;
            dance[2].y -= 95/speed;
            dance[2].rotate -= 90/speed;
            dance[1].x += 35/speed;
            dance[1].y -= 15/speed;
            dance[1].rotate -= 75/speed;
            dance[8].x += 50/speed;
            dance[8].y -= 60/speed;
            dance[8].rotate -= 10/speed;
            dance[7].x += 20/speed;
            dance[7].y -= 25/speed;
            dance[7].rotate -= 70/speed;
            //Utstilling 2
            bop[0].y1 += 50/speed;
            bop[0].y2 += 50/speed;
            bop[1].y += 50/speed;
            bop[2].y += 50/speed;
            bop[3].y += 50/speed;
            bop[4].y += 50/speed;
            bop[9].y += 50/speed;
            bop[10].y += 50/speed;
            bop[6].rotate -= 20/speed;
            bop[8].rotate += 20/speed;
            bop[5].rotate += 20/speed;
            bop[7].rotate -= 20/speed;
            j++;
        } else {
            //Utstilling 1
            dance[2].x -= 40/speed;
            dance[2].y += 95/speed;
            dance[2].rotate += 90/speed;
            dance[1].x -= 35/speed;
            dance[1].y += 15/speed;
            dance[1].rotate += 75/speed;
            dance[8].x -= 50/speed;
            dance[8].y += 60/speed;
            dance[8].rotate += 10/speed;
            dance[7].x -= 20/speed;
            dance[7].y += 25/speed;
            dance[7].rotate += 70/speed;
            //Utstilling 2
            bop[0].y1 -= 50/speed;
            bop[0].y2 -= 50/speed;
            bop[1].y -= 50/speed;
            bop[2].y -= 50/speed;
            bop[3].y -= 50/speed;
            bop[4].y -= 50/speed;
            bop[9].y -= 50/speed;
            bop[10].y -= 50/speed;
            bop[6].rotate += 20/speed;
            bop[8].rotate -= 20/speed;
            bop[5].rotate -= 20/speed;
            bop[7].rotate += 20/speed;
            j--;
        }
        if(j > speed){
            forward = false;
        } else if (j<1) {
            forward = true;
        }
        requestAnimationFrame(animate);
    }
});

