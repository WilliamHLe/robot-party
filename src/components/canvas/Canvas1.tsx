import React, { useRef, useEffect} from 'react'
import "../container.css"

interface CanvasProps {
    speedValue: number;
    colorValue: string;
}

const Canvas1 = ({speedValue, colorValue}: CanvasProps, props: JSX.IntrinsicAttributes & React.ClassAttributes<HTMLCanvasElement> & React.CanvasHTMLAttributes<HTMLCanvasElement>) => {

    const canvasRef = useRef(null)

    let j = 1;
    let forward = true;
    //speedValue oppdateres til riktig verdi, men vises ikke i bildet?
    //problem med render?
    let speed = speedValue;
    //let utstilling = "bop";
    let dance = [
        //Head
        { x: 320, y: 175, w: 105, h: 125, rotate: 0 },
        //Left arm
        { x: 365, y: 290, w: 50, h: 100, rotate: -10 },
        {
            x: 385,
            y: 355,
            w: 100,
            h: 50,
            rotate: 10
        },
        //Left foot
        { x: 400, y: 400, w: 60, h: 125, rotate: -40 },
        { x: 425, y: 490, w: 50, h: 100, rotate: 10 },
        //Body
        { x: 375, y: 325, w: 0, h: 0, rotate: 0 },
        { x: 375, y: 375, w: 0, h: 0, rotate: 0 },
        //Right arm
        { x: 345, y: 300, w: 50, h: 100, rotate: -10 },
        {
            x: 365,
            y: 365,
            w: 100,
            h: 50,
            rotate: 10
        },
        //Right foot
        { x: 325, y: 415, w: 60, h: 125, rotate: 5 },
        { x: 315, y: 515, w: 50, h: 100, rotate: 20 }
    ];

    const drawFloor = (ctx:any) => {
        ctx.globalAlpha = 1;
        let gradient = ctx.createLinearGradient(0, 450, 0, 750);
        gradient.addColorStop(0, "grey");
        gradient.addColorStop(1, "white");
        ctx.fillStyle = gradient;
        ctx.fillRect(0, 0, 800, 800);
        //ctx.globalAlpha = 0.5;
    }

    const drawDancer = (arr:any, ctx:any) => {
        arr.forEach((r: { w: number; x: number; y: number; x1: any; y1: any; x2: any; y2: any; x3: any; h: number; rotate: number; }) => {
            ctx.save();
            ctx.beginPath();
            ctx.fillStyle = "yellow";
            if (r.w === 0) {
                ctx.arc(r.x, r.y, 55, 0, 2 * Math.PI);
            } else if (typeof r.x1 !== "undefined") {
                ctx.beginPath();
                ctx.moveTo(r.x1, r.y1);
                ctx.lineTo(r.x2, r.y2);
                ctx.lineTo(r.x3, r.y2);
                ctx.closePath();
                ctx.stroke();
                ctx.fill();
            } else {
                //Kode som gjÃ¸r av rektangler roterer rundt midtpunket sitt. Koden er funnet her: https://stackoverflow.com/questions/17125632/html5-canvas-rotate-object-without-moving-coordinates/17126036
                ctx.translate(r.x + r.w / 2, r.y + r.h / 2);
                ctx.rotate((r.rotate * Math.PI) / 180);
                ctx.rect(-r.w / 2, -r.h / 2, r.w, r.h);
            }
            ctx.fill();
            ctx.stroke();
            ctx.restore();
        });
    }

    useEffect(() => {
        const canvas:any = canvasRef.current
        const context = canvas.getContext('2d')

        const animate = () => {
            context.clearRect(0, 0, 800, 800);
            drawFloor(context)
            drawDancer(dance, context);
            if(forward){
                dance[2].x += 40/speed;
                dance[2].y -= 95/speed;
                dance[2].rotate -= 90/speed;
                dance[1].x += 35/speed;
                dance[1].y -= 15/speed;
                dance[1].rotate -= 75/speed;
                dance[8].x += 50/speed;
                dance[8].y -= 60/speed;
                dance[8].rotate -= 10/speed;
                dance[7].x += 20/speed;
                dance[7].y -= 25/speed;
                dance[7].rotate -= 70/speed;
                j++;
            } else {
                dance[2].x -= 40/speed;
                dance[2].y += 95/speed;
                dance[2].rotate += 90/speed;
                dance[1].x -= 35/speed;
                dance[1].y += 15/speed;
                dance[1].rotate += 75/speed;
                dance[8].x -= 50/speed;
                dance[8].y += 60/speed;
                dance[8].rotate += 10/speed;
                dance[7].x -= 20/speed;
                dance[7].y += 25/speed;
                dance[7].rotate += 70/speed;
                j--;
            }
            if(j > speed){
                forward = false;
            } else if (j<1) {
                forward = true;
            }
            requestAnimationFrame(animate);
        }
        requestAnimationFrame(animate);
    }, [])

    return <canvas ref={canvasRef} width={800} height={800} className="canvas"  {...props}/>
}

export default Canvas1