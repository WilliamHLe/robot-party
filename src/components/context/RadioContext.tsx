import {createContext} from "react"

const RadioContext = createContext({
    speed: 60,
    color: "#FF0000"
})

export default RadioContext;