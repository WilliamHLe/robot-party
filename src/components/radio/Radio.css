.radio-wrapper {
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;
}
.radio-element {
    display: flex;
    flex-flow: column;
    align-items: center;
    padding: 30px;
}