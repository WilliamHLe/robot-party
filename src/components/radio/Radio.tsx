import React, {useState} from "react";
import "./Radio.css";
import Carousel from "../carousel/Carousel";
import Slide from "../carousel/Slide";
import RadioContext from "../context/RadioContext";

import Canvas1 from "../canvas/Canvas1";
import Canvas2 from "../canvas/Canvas2";
import Canvas3 from "../canvas/Canvas3";
import Canvas4 from "../canvas/Canvas4";
import Canvas5 from "../canvas/Canvas5";

// To radioknapper
// En for hastighet (lagres som tall)
// En for farger (hex-verdier, lagres som streng)


const Radio: React.FC = () => {

    //60 = SAKTE ---- 35 = NORMAL ---- 15 = RASK
    const [speed, setSpeed] = useState<number>(60);

    // "#FF0000" = RØDT ---- "#0000FF" ---- BLÅTT "#00FF00" ---- GRØNT
    const [color, setColor] = useState<string>("#FF0000");


    return(
        <div>
            <div className="radio-wrapper">
                <div className="radio-element">
                    <h3>Velg hastighet:</h3>
                    <div>
                        <label>
                            <input type="radio" value={60} name="Sakte" checked={speed === 60} onChange={() => setSpeed(60)}/>
                            Sakte&nbsp;&nbsp;
                        </label>
                        <label>
                            <input type="radio" value={35} name="Normal" checked={speed === 35} onChange={() => setSpeed(35)}/>
                            Normal&nbsp;&nbsp;
                        </label>
                        <label>
                            <input type="radio" value={15} name="Rask" checked={speed === 15} onChange={() => setSpeed(15)}/>
                            Rask
                        </label>

                    </div>
                </div>
                <div className="radio-element">
                    <h3>Velg fargetema:</h3>
                    <div>

                        <label>
                            <input type="radio" value="#FF0000" name="Rødt" checked={color === "#FF0000"} onChange={() => setColor("#FF0000")}/>
                            Rødt&nbsp;&nbsp;
                        </label>
                        <label>
                            <input type="radio" value="#0000FF" name="Blått" checked={color === "#0000FF"} onChange={() => setColor("#0000FF")}/>
                            Blått&nbsp;&nbsp;
                        </label>
                        <label>
                            <input type="radio" value="#00FF00" name="Grønt" checked={color === "#00FF00"} onChange={() => setColor("#00FF00")}/>
                            Grønt
                        </label>
                    </div>
                </div>
            </div>
            {/*Bruker RadioContext.Provider til å sende states for "speed" og "color" nedover i hierarkiet (til Canvas-komponentene)*/}
            <RadioContext.Provider value={{speed, color}}>
                <div>
                    <Carousel>
                        <Slide url = {"https://poetrydb.org/author,title/Emily%20Dickinson;We%20should%20not%20mind%20so%20small%20a%20flower"} canvas = {<Canvas1/>}/>
                        <Slide url = {"https://poetrydb.org/author,title/Emily%20Dickinson;Could%20Hope%20inspect%20her%20Basis"} canvas = {<Canvas2/>}/>
                        <Slide url = {"https://poetrydb.org/author,title/Emily%20Dickinson;Success%20is%20counted%20sweetest"} canvas = {<Canvas3/>}/>
                        <Slide url = {"https://poetrydb.org/author,title/Emily%20Dickinson;Nature%20can%20do%20no%20more"} canvas = {<Canvas4/>}/>
                        <Slide url = {"https://poetrydb.org/author,title/Emily%20Dickinson;The%20Butterfly's%20Numidian%20Gown"} canvas = {<Canvas5/>}/>
                    </Carousel>
                </div>
            </RadioContext.Provider>
        </div>
    );

};

export default Radio;