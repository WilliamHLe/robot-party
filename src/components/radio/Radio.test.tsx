import React from 'react';
import Radio from "./Radio";
/*
import RadioContext from "../context/RadioContext";
import {act, render, fireEvent, cleanup} from '@testing-library/react';
import Canvas1 from "../canvas/Canvas1";
*/
const renderer = require("react-test-renderer");


it('renders correctly', () => {
    const tree = renderer.create(<Radio />).toJSON();
    expect(tree).toMatchSnapshot();
});

//test context??


/*
afterEach(cleanup)

it('Context value is updated by child component', () => {

    const { container, getByText } = render(<Radio/>);

    expect(getByText(/Some/i).textContent).toBe("Some Text")

    fireEvent.click(getByText("Change Text"))

    expect(getByText(/Some/i).textContent).toBe("Some Other Text")
})
*/
