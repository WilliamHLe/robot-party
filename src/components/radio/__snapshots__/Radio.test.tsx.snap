// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    className="radio-wrapper"
  >
    <div
      className="radio-element"
    >
      <h3>
        Velg hastighet:
      </h3>
      <div>
        <label>
          <input
            checked={true}
            name="Sakte"
            onChange={[Function]}
            type="radio"
            value={60}
          />
          Sakte  
        </label>
        <label>
          <input
            checked={false}
            name="Normal"
            onChange={[Function]}
            type="radio"
            value={35}
          />
          Normal  
        </label>
        <label>
          <input
            checked={false}
            name="Rask"
            onChange={[Function]}
            type="radio"
            value={15}
          />
          Rask
        </label>
      </div>
    </div>
    <div
      className="radio-element"
    >
      <h3>
        Velg fargetema:
      </h3>
      <div>
        <label>
          <input
            checked={true}
            name="Rødt"
            onChange={[Function]}
            type="radio"
            value="#FF0000"
          />
          Rødt  
        </label>
        <label>
          <input
            checked={false}
            name="Blått"
            onChange={[Function]}
            type="radio"
            value="#0000FF"
          />
          Blått  
        </label>
        <label>
          <input
            checked={false}
            name="Grønt"
            onChange={[Function]}
            type="radio"
            value="#00FF00"
          />
          Grønt
        </label>
      </div>
    </div>
    <div
      className="radio-element"
    >
      <h3>
        Velg musikk:
      </h3>
      <div>
        <label>
          <input
            checked={true}
            name="musicChoice1"
            onChange={[Function]}
            type="radio"
            value="bensoundmoose.mp3"
          />
           Musikk 1
        </label>
        <label>
          <input
            checked={false}
            name="musicChoice2"
            onChange={[Function]}
            type="radio"
            value="ColourOfYourFaceNoMBe.mp3"
          />
           Musikk 2
        </label>
        <label>
          <input
            checked={false}
            name="musicChoice3"
            onChange={[Function]}
            type="radio"
            value="RobotsaComethDanLebowitz.mp3"
          />
           Musikk 3
        </label>
        <br />
        <audio
          autoPlay={true}
          id="musicPlayer"
          onLoad={[Function]}
        >
          <source
            src="bensoundmoose.mp3"
            type="audio/mpeg"
          />
        </audio>
        <span
          className="pause"
          id="pButton"
          onClick={[Function]}
        >
           
        </span>
        <span
          className="lowVolume"
        >
           
        </span>
        <input
          className="slider"
          defaultValue="50"
          id="volumeSlider"
          max="100"
          min="0"
          onChange={[Function]}
          type="range"
        />
        <span
          className="maxVolume"
        >
           
        </span>
      </div>
    </div>
  </div>
  <div>
    <div>
      <div
        className="carousel-wrapper"
      >
        <div
          className="carousel-slides"
        >
          <div
            className="slide-active"
          >
            <div
              className="grid-container"
            >
              <div
                className="grid-canvas"
              >
                <canvas
                  className="canvas"
                  height={800}
                  width={800}
                />
              </div>
              <div
                className="grid-poetry"
              >
                <div
                  className="poetry-container"
                >
                  <h3>
                    
                     av 
                    
                  </h3>
                  <div>
                    <p>
                      
                    </p>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            className="slide-hidden"
          >
            <div
              className="grid-container"
            >
              <div
                className="grid-canvas"
              >
                <canvas
                  className="canvas"
                  height={800}
                  width={800}
                />
              </div>
              <div
                className="grid-poetry"
              >
                <div
                  className="poetry-container"
                >
                  <h3>
                    
                     av 
                    
                  </h3>
                  <div>
                    <p>
                      
                    </p>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            className="slide-hidden"
          >
            <div
              className="grid-container"
            >
              <div
                className="grid-canvas"
              >
                <canvas
                  className="canvas"
                  height={800}
                  width={800}
                />
              </div>
              <div
                className="grid-poetry"
              >
                <div
                  className="poetry-container"
                >
                  <h3>
                    
                     av 
                    
                  </h3>
                  <div>
                    <p>
                      
                    </p>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            className="slide-hidden"
          >
            <div
              className="grid-container"
            >
              <div
                className="grid-canvas"
              >
                <canvas
                  className="canvas"
                  height={800}
                  width={800}
                />
              </div>
              <div
                className="grid-poetry"
              >
                <div
                  className="poetry-container"
                >
                  <h3>
                    
                     av 
                    
                  </h3>
                  <div>
                    <p>
                      
                    </p>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            className="slide-hidden"
          >
            <div
              className="grid-container"
            >
              <div
                className="grid-canvas"
              >
                <canvas
                  className="canvas"
                  height={800}
                  width={800}
                />
              </div>
              <div
                className="grid-poetry"
              >
                <div
                  className="poetry-container"
                >
                  <h3>
                    
                     av 
                    
                  </h3>
                  <div>
                    <p>
                      
                    </p>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="carousel-buttons"
        >
          <button
            className="carousel-button"
            onClick={[Function]}
            style={
              Object {
                "float": "left",
              }
            }
          >
            ≪ Forrige
          </button>
          <button
            className="carousel-button"
            onClick={[Function]}
            style={
              Object {
                "float": "right",
              }
            }
          >
            Neste ≫
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
